test/new-e2e/pkg/utils/e2e/stack_definition.go

```diff
@@ -72,29 +73,98 @@ type AgentEnv struct {
 	Agent *client.Agent
 }
 
+// AgentStackDefParam defines the parameters for a stack with a VM and the Datadog Agent
+// installed.
+// The AgentStackDefParam configuration uses the [Functional options pattern].
+//
+// The available options are:
+//   - [WithVMParams]
+//   - [WithAgentParams]
+//   - [WithAgentClientParams]
+//
+// [Functional options pattern]: https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis
+type AgentStackDefParam struct {
+	vmParams          []ec2params.Option
+	agentClientParams []agentclientparams.Option
+	agentParams       []agentparams.Option
+}
+
+func newAgentStackDefParam(options ...func(*AgentStackDefParam) error) (*AgentStackDefParam, error) {
+	params := &AgentStackDefParam{
+		vmParams:          []ec2params.Option{},
+		agentClientParams: []agentclientparams.Option{},
+		agentParams:       []agentparams.Option{},
+	}
+	for _, o := range options {
+		err := o(params)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return params, nil
+}
+
+// WithVMParams sets VM parameters
+// See [ec2vm.Params] for available options for vmParams.
+func WithVMParams(options ...ec2params.Option) func(*AgentStackDefParam) error {
+	return func(p *AgentStackDefParam) error {
+		p.vmParams = options
+		return nil
+	}
+}
+
+// WithAgentParams sets Agent parameters
+// See [agent.Params] for available options for agentParams.
+func WithAgentParams(options ...agentparams.Option) func(*AgentStackDefParam) error {
+	return func(p *AgentStackDefParam) error {
+		p.agentParams = options
+		return nil
+	}
+}
+
+// WithAgentClientParams sets Agent client parameters
+// See [agentclientparams.Params] for available options for agentParams
+func WithAgentClientParams(options ...agentclientparams.Option) func(*AgentStackDefParam) error {
+	return func(p *AgentStackDefParam) error {
+		p.agentClientParams = options
+		return nil
+	}
+}
+
 // AgentStackDef creates a stack definition containing a virtual machine and an Agent.
 //
 // See [ec2vm.Params] for available options for vmParams.
 //
 // See [agent.Params] for available options for agentParams.
 //
+// See [agentclientparams.Params] for available options for agentParams
+//
 // [ec2vm.Params]: https://pkg.go.dev/github.com/DataDog/test-infra-definitions@main/scenarios/aws/vm/ec2VM#Params
 // [agent.Params]: https://pkg.go.dev/github.com/DataDog/test-infra-definitions@main/components/datadog/agent#Params
-func AgentStackDef(vmParams []ec2params.Option, agentParameters ...agentparams.Option) *StackDefinition[AgentEnv] {
+// [agentclientparams.Params]: https://pkg.go.dev/github.com/DataDog/datadog-agent@main/test/new-e2e/pkg/utils/e2e/client/agentclientparams#Params
+func AgentStackDef(options ...func(*AgentStackDefParam) error) *StackDefinition[AgentEnv] {
```

Comment:
For reviewers: this is the biggest change: adding functional options to stack definitions. I used one single param struct for both for simplicity, I can move it to a dedicated package.